Submitted By: Ken Moffat <ken at linuxfromscratch dot org>
Date: 2017-09-21
Initial Package Version: 20170524
Upstream Status: Unknown
Origin: Arch linux
Description: Fixes compilation with the API changes of poppler-0.59.

diff -ur source.orig/texk/web2c/pdftexdir/pdftoepdf.cc source/texk/web2c/pdftexdir/pdftoepdf.cc
--- source.orig/texk/web2c/pdftexdir/pdftoepdf.cc	2017-06-16 21:49:09.293857387 +0200
+++ source/texk/web2c/pdftexdir/pdftoepdf.cc	2017-09-19 09:48:19.661355046 +0200
@@ -84,31 +84,6 @@
 #define MASK_SUPPRESS_PTEX_PAGENUMBER 0x04
 #define MASK_SUPPRESS_PTEX_INFODICT   0x08
 
-// PdfObject encapsulates the xpdf Object type,
-// and properly frees its resources on destruction.
-// Use obj-> to access members of the Object,
-// and &obj to get a pointer to the object.
-// It is no longer necessary to call Object::free explicitely.
-
-class PdfObject {
-  public:
-    PdfObject() {               // nothing
-    } ~PdfObject() {
-        iObject.free();
-    }
-    Object *operator->() {
-        return &iObject;
-    }
-    Object *operator&() {
-        return &iObject;
-    }
-  private:                     // no copying or assigning
-    PdfObject(const PdfObject &);
-    void operator=(const PdfObject &);
-  public:
-    Object iObject;
-};
-
 // When copying the Resources of the selected page, all objects are copied
 // recusively top-down. Indirect objects however are not fetched during
 // copying, but get a new object number from pdfTeX and then will be
@@ -212,18 +187,6 @@
     delete pdf_doc;
 }
 
-// Replacement for
-//      Object *initDict(Dict *dict1){ initObj(objDict); dict = dict1; return this; }
-
-static void initDictFromDict(PdfObject & obj, Dict * dict)
-{
-    obj->initDict(xref);
-    for (int i = 0, l = dict->getLength(); i < l; i++) {
-        Object obj1;
-        obj->dictAdd(copyString(dict->getKey(i)), dict->getValNF(i, &obj1));
-    }
-}
-
 // --------------------------------------------------------------------
 
 static int addEncoding(GfxFont * gfont)
@@ -320,10 +283,10 @@
 
 static void copyDictEntry(Object * obj, int i)
 {
-    PdfObject obj1;
+    Object obj1;
     copyName(obj->dictGetKey(i));
     pdf_puts(" ");
-    obj->dictGetValNF(i, &obj1);
+    obj1 = obj->dictGetValNF(i);
     copyObject(&obj1);
     pdf_puts("\n");
 }
@@ -376,17 +339,17 @@
 static void copyProcSet(Object * obj)
 {
     int i, l;
-    PdfObject procset;
+    Object procset;
     if (!obj->isArray())
         pdftex_fail("PDF inclusion: invalid ProcSet array type <%s>",
                     obj->getTypeName());
     pdf_puts("/ProcSet [ ");
     for (i = 0, l = obj->arrayGetLength(); i < l; ++i) {
-        obj->arrayGetNF(i, &procset);
-        if (!procset->isName())
+        procset = obj->arrayGetNF(i);
+        if (!procset.isName())
             pdftex_fail("PDF inclusion: invalid ProcSet entry type <%s>",
-                        procset->getTypeName());
-        copyName(procset->getName());
+                        procset.getTypeName());
+        copyName(procset.getName());
         pdf_puts(" ");
     }
     pdf_puts("]\n");
@@ -396,7 +359,7 @@
 
 static void copyFont(char *tag, Object * fontRef)
 {
-    PdfObject fontdict, subtype, basefont, fontdescRef, fontdesc, charset,
+    Object fontdict, subtype, basefont, fontdescRef, fontdesc, charset,
         fontfile, ffsubtype, stemV;
     GfxFont *gfont;
     fd_entry *fd;
@@ -413,33 +376,49 @@
     }
     // Only handle included Type1 (and Type1C) fonts; anything else will be copied.
     // Type1C fonts are replaced by Type1 fonts, if REPLACE_TYPE1C is true.
-    if (!fixedinclusioncopyfont && fontRef->fetch(xref, &fontdict)->isDict()
-        && fontdict->dictLookup("Subtype", &subtype)->isName()
-        && !strcmp(subtype->getName(), "Type1")
-        && fontdict->dictLookup("BaseFont", &basefont)->isName()
-        && fontdict->dictLookupNF("FontDescriptor", &fontdescRef)->isRef()
-        && fontdescRef->fetch(xref, &fontdesc)->isDict()
-        && (fontdesc->dictLookup("FontFile", &fontfile)->isStream()
-            || (REPLACE_TYPE1C
-                && fontdesc->dictLookup("FontFile3", &fontfile)->isStream()
-                && fontfile->streamGetDict()->lookup("Subtype",
-                                                     &ffsubtype)->isName()
-                && !strcmp(ffsubtype->getName(), "Type1C")))
-        && (fontmap = lookup_fontmap(basefont->getName())) != NULL) {
+    fontdict = fontRef->fetch(xref);
+    if (fontdict.isDict()) {
+        subtype = fontdict.dictLookup("Subtype");
+        basefont = fontdict.dictLookup("BaseFont");
+        fontdescRef = fontdict.dictLookupNF("FontDescriptor");
+        if (fontdescRef.isRef()) {
+            fontdesc = fontdescRef.fetch(xref);
+            if (fontdesc.isDict()) {
+                fontfile = fontdesc.dictLookup("FontFile");
+                if (!fontfile.isStream() && REPLACE_TYPE1C) {
+                    fontfile = fontdesc.dictLookup("FontFile3");
+                    ffsubtype = fontfile.streamGetDict()->lookup("Subtype");
+                    if (!(ffsubtype.isName() && !strcmp(ffsubtype.getName(), "Type1C"))) {
+                        // not a Type1-C font.
+                        fontfile = Object(objNull);
+                    }
+                }
+            }
+        }
+    }
+    if (!fixedinclusioncopyfont && fontdict.isDict()
+        && subtype.isName()
+        && !strcmp(subtype.getName(), "Type1")
+        && basefont.isName()
+        && fontdescRef.isRef()
+        && fontdesc.isDict()
+        && fontfile.isStream()
+        && (fontmap = lookup_fontmap(basefont.getName())) != NULL) {
         // round /StemV value, since the PDF input is a float
         // (see Font Descriptors in PDF reference), but we only store an
         // integer, since we don't want to change the struct.
-        fontdesc->dictLookup("StemV", &stemV);
-        fd = epdf_create_fontdescriptor(fontmap, zround(stemV->getNum()));
-        if (fontdesc->dictLookup("CharSet", &charset) &&
-            charset->isString() && is_subsetable(fontmap))
-            epdf_mark_glyphs(fd, charset->getString()->getCString());
+        stemV = fontdesc.dictLookup("StemV");
+        fd = epdf_create_fontdescriptor(fontmap, zround(stemV.getNum()));
+        charset = fontdesc.dictLookup("CharSet");
+        if (!charset.isNull() &&
+            charset.isString() && is_subsetable(fontmap))
+            epdf_mark_glyphs(fd, charset.getString()->getCString());
         else
             embed_whole_font(fd);
-        addFontDesc(fontdescRef->getRef(), fd);
+        addFontDesc(fontdescRef.getRef(), fd);
         copyName(tag);
         gfont = GfxFont::makeFont(xref, tag, fontRef->getRef(),
-                                  fontdict->getDict());
+                                  fontdict.getDict());
         pdf_printf(" %d 0 R ", addFont(fontRef->getRef(), fd,
                                        addEncoding(gfont)));
     } else {
@@ -451,24 +430,24 @@
 
 static void copyFontResources(Object * obj)
 {
-    PdfObject fontRef;
+    Object fontRef;
     int i, l;
     if (!obj->isDict())
         pdftex_fail("PDF inclusion: invalid font resources dict type <%s>",
                     obj->getTypeName());
     pdf_puts("/Font << ");
     for (i = 0, l = obj->dictGetLength(); i < l; ++i) {
-        obj->dictGetValNF(i, &fontRef);
-        if (fontRef->isRef())
+        fontRef = obj->dictGetValNF(i);
+        if (fontRef.isRef())
             copyFont(obj->dictGetKey(i), &fontRef);
-        else if (fontRef->isDict()) {   // some programs generate pdf with embedded font object
+        else if (fontRef.isDict()) {   // some programs generate pdf with embedded font object
             copyName(obj->dictGetKey(i));
             pdf_puts(" ");
             copyObject(&fontRef);
         }
         else
             pdftex_fail("PDF inclusion: invalid font in reference type <%s>",
-                        fontRef->getTypeName());
+                        fontRef.getTypeName());
     }
     pdf_puts(">>\n");
 }
@@ -557,7 +536,7 @@
 
 static void copyObject(Object * obj)
 {
-    PdfObject obj1;
+    Object obj1;
     int i, l, c;
     Ref ref;
     char *p;
@@ -601,8 +580,8 @@
     } else if (obj->isArray()) {
         pdf_puts("[");
         for (i = 0, l = obj->arrayGetLength(); i < l; ++i) {
-            obj->arrayGetNF(i, &obj1);
-            if (!obj1->isName())
+            obj1 = obj->arrayGetNF(i);
+            if (!obj1.isName())
                 pdf_puts(" ");
             copyObject(&obj1);
         }
@@ -612,7 +591,7 @@
         copyDict(obj);
         pdf_puts(">>");
     } else if (obj->isStream()) {
-        initDictFromDict(obj1, obj->streamGetDict());
+        obj1 = Object(obj->streamGetDict());
         pdf_puts("<<\n");
         copyDict(&obj1);
         pdf_puts(">>\n");
@@ -638,9 +617,8 @@
     InObj *r;
     for (r = inObjList; r != 0; r = r->next) {
         if (!r->written) {
-            Object obj1;
             r->written = 1;
-            xref->fetch(r->ref.num, r->ref.gen, &obj1);
+            Object obj1 = xref->fetch(r->ref.num, r->ref.gen);
             if (r->type == objFont) {
                 assert(!obj1.isStream());
                 pdfbeginobj(r->num, 2);         // \pdfobjcompresslevel = 2 is for this
@@ -656,7 +634,6 @@
                 pdf_puts("\n");
                 pdfendobj();
             }
-            obj1.free();
         }
     }
 }
@@ -839,8 +816,8 @@
     Page *page;
     Ref *pageRef;
     Dict *pageDict;
-    PdfObject contents, obj1, obj2, pageObj, dictObj;
-    PdfObject groupDict;
+    Object contents, obj1, obj2, pageObj, dictObj;
+    Object groupDict;
     bool writeSepGroup = false;
     Object info;
     char *key;
@@ -867,8 +844,8 @@
     encodingList = 0;
     page = pdf_doc->doc->getCatalog()->getPage(epdf_selected_page);
     pageRef = pdf_doc->doc->getCatalog()->getPageRef(epdf_selected_page);
-    xref->fetch(pageRef->num, pageRef->gen, &pageObj);
-    pageDict = pageObj->getDict();
+    pageObj = xref->fetch(pageRef->num, pageRef->gen);
+    pageDict = pageObj.getDict();
     rotate = page->getRotate();
     PDFRectangle *pagebox;
     // write the Page header
@@ -886,7 +863,7 @@
         pdf_printf("/%s.PageNumber %i\n", pdfkeyprefix, (int) epdf_selected_page);
     }
     if ((suppress_ptex_info & MASK_SUPPRESS_PTEX_INFODICT) == 0) {
-        pdf_doc->doc->getDocInfoNF(&info);
+        info = pdf_doc->doc->getDocInfoNF();
         if (info.isRef()) {
             // the info dict must be indirect (PDF Ref p. 61)
             pdf_printf("/%s.InfoDict ", pdfkeyprefix);
@@ -942,14 +919,14 @@
     pdf_puts(stripzeros(s));
 
     // Metadata validity check (as a stream it must be indirect)
-    pageDict->lookupNF("Metadata", &dictObj);
-    if (!dictObj->isNull() && !dictObj->isRef())
+    dictObj = pageDict->lookupNF("Metadata");
+    if (!dictObj.isNull() && !dictObj.isRef())
         pdftex_warn("PDF inclusion: /Metadata must be indirect object");
 
     // copy selected items in Page dictionary except Resources & Group
     for (i = 0; pageDictKeys[i] != NULL; i++) {
-        pageDict->lookupNF(pageDictKeys[i], &dictObj);
-        if (!dictObj->isNull()) {
+        dictObj = pageDict->lookupNF(pageDictKeys[i]);
+        if (!dictObj.isNull()) {
             pdf_newline();
             pdf_printf("/%s ", pageDictKeys[i]);
             copyObject(&dictObj); // preserves indirection
@@ -957,8 +934,8 @@
     } 
 
     // handle page group
-    pageDict->lookupNF("Group", &dictObj);
-    if (!dictObj->isNull()) {
+    dictObj = pageDict->lookupNF("Group");
+    if (!dictObj.isNull()) {
         if (pdfpagegroupval == 0) { 
             // another pdf with page group was included earlier on the
             // same page; copy the Group entry as is.  See manual for
@@ -972,11 +949,11 @@
             copyObject(&dictObj);
         } else {
             // write Group dict as a separate object, since the Page dict also refers to it
-            pageDict->lookup("Group", &dictObj);
-            if (!dictObj->isDict())
+            dictObj = pageDict->lookup("Group");
+            if (!dictObj.isDict())
                 pdftex_fail("PDF inclusion: /Group dict missing");
             writeSepGroup = true;
-            initDictFromDict(groupDict, page->getGroup());
+            groupDict = Object(page->getGroup());
             pdf_printf("/Group %ld 0 R\n", (long)pdfpagegroupval);
         }
     }
@@ -989,15 +966,15 @@
         pdftex_warn
             ("PDF inclusion: /Resources missing. 'This practice is not recommended' (PDF Ref)");
     } else {
-        initDictFromDict(obj1, page->getResourceDict());
-        if (!obj1->isDict())
+        obj1 = Object(page->getResourceDict());
+        if (!obj1.isDict())
             pdftex_fail("PDF inclusion: invalid resources dict type <%s>",
-                        obj1->getTypeName());
+                        obj1.getTypeName());
         pdf_newline();
         pdf_puts("/Resources <<\n");
-        for (i = 0, l = obj1->dictGetLength(); i < l; ++i) {
-            obj1->dictGetVal(i, &obj2);
-            key = obj1->dictGetKey(i);
+        for (i = 0, l = obj1.dictGetLength(); i < l; ++i) {
+            obj2 = obj1.dictGetVal(i);
+            key = obj1.dictGetKey(i);
             if (strcmp("Font", key) == 0)
                 copyFontResources(&obj2);
             else if (strcmp("ProcSet", key) == 0)
@@ -1009,8 +986,8 @@
     }
 
     // write the page contents
-    page->getContents(&contents);
-    if (contents->isStream()) {
+    contents = page->getContents();
+    if (contents.isStream()) {
 
         // Variant A: get stream and recompress under control
         // of \pdfcompresslevel
@@ -1021,36 +998,35 @@
 
         // Variant B: copy stream without recompressing
         //
-        contents->streamGetDict()->lookup("F", &obj1);
-        if (!obj1->isNull()) {
+        obj1 = contents.streamGetDict()->lookup("F");
+        if (!obj1.isNull()) {
             pdftex_fail("PDF inclusion: Unsupported external stream");
         }
-        contents->streamGetDict()->lookup("Length", &obj1);
-        assert(!obj1->isNull());
+        obj1 = contents.streamGetDict()->lookup("Length");
+        assert(!obj1.isNull());
         pdf_puts("/Length ");
         copyObject(&obj1);
         pdf_puts("\n");
-        contents->streamGetDict()->lookup("Filter", &obj1);
-        if (!obj1->isNull()) {
+        obj1 = contents.streamGetDict()->lookup("Filter");
+        if (!obj1.isNull()) {
             pdf_puts("/Filter ");
             copyObject(&obj1);
             pdf_puts("\n");
-            contents->streamGetDict()->lookup("DecodeParms", &obj1);
-            if (!obj1->isNull()) {
+            obj1 = contents.streamGetDict()->lookup("DecodeParms");
+            if (!obj1.isNull()) {
                 pdf_puts("/DecodeParms ");
                 copyObject(&obj1);
                 pdf_puts("\n");
             }
         }
         pdf_puts(">>\nstream\n");
-        copyStream(contents->getStream()->getUndecodedStream());
+        copyStream(contents.getStream()->getUndecodedStream());
         pdfendstream();
-    } else if (contents->isArray()) {
+    } else if (contents.isArray()) {
         pdfbeginstream();
-        for (i = 0, l = contents->arrayGetLength(); i < l; ++i) {
-            Object contentsobj;
-            copyStream((contents->arrayGet(i, &contentsobj))->getStream());
-            contentsobj.free();
+        for (i = 0, l = contents.arrayGetLength(); i < l; ++i) {
+            Object contentsobj = contents.arrayGet(i);
+            copyStream(contentsobj.getStream());
             if (i < l - 1)
                 pdf_newline();  // add a newline after each stream except the last
         }
diff -ur source.orig/texk/web2c/pdftexdir/pdftosrc.cc source/texk/web2c/pdftexdir/pdftosrc.cc
--- source.orig/texk/web2c/pdftexdir/pdftosrc.cc	2017-06-16 21:49:09.297190679 +0200
+++ source/texk/web2c/pdftexdir/pdftosrc.cc	2017-09-19 09:48:19.661355046 +0200
@@ -86,22 +86,20 @@
             objgen = atoi(argv[3]);
     }
     xref = doc->getXRef();
-    catalogDict.initNull();
-    xref->getCatalog(&catalogDict);
+    catalogDict = xref->getCatalog();
     if (!catalogDict.isDict("Catalog")) {
         fprintf(stderr, "No Catalog found\n");
         exit(1);
     }
-    srcStream.initNull();
+    srcStream = Object(objNull);
     if (objnum == 0) {
-        catalogDict.dictLookup("SourceObject", &srcStream);
+        srcStream = catalogDict.dictLookup("SourceObject");
         static char const_SourceFile[] = "SourceFile";
         if (!srcStream.isStream(const_SourceFile)) {
             fprintf(stderr, "No SourceObject found\n");
             exit(1);
         }
-        srcName.initNull();
-        srcStream.getStream()->getDict()->lookup("SourceName", &srcName);
+        srcName = srcStream.getStream()->getDict()->lookup("SourceName");
         if (!srcName.isString()) {
             fprintf(stderr, "No SourceName found\n");
             exit(1);
@@ -110,7 +108,7 @@
         // We cannot free srcName, as objname shares its string.
         // srcName.free();
     } else if (objnum > 0) {
-        xref->fetch(objnum, objgen, &srcStream);
+        srcStream = xref->fetch(objnum, objgen);
         if (!srcStream.isStream()) {
             fprintf(stderr, "Not a Stream object\n");
             exit(1);
@@ -160,26 +158,24 @@
                 int localOffset = 0;
                 Guint firstOffset;
 
-                assert(xref->fetch(e->offset, 0, &objStr)->isStream());
-                nObjects = objStr.streamGetDict()->lookup("N", &obj1)->getInt();
-                obj1.free();
-                first = objStr.streamGetDict()->lookup("First", &obj1)->getInt();
-                obj1.free();
+                objStr = xref->fetch(e->offset, 0);
+                assert(objStr.isStream());
+                obj1 = objStr.streamGetDict()->lookup("N");
+                nObjects = obj1.getInt();
+                obj1 = objStr.streamGetDict()->lookup("First");
+                first = obj1.getInt();
                 firstOffset = objStr.getStream()->getBaseStream()->getStart() + first;
 
                 // parse the header: object numbers and offsets
                 objStr.streamReset();
-                obj1.initNull();
-                str = new EmbedStream(objStr.getStream(), &obj1, gTrue, first);
+                str = new EmbedStream(objStr.getStream(), Object(objNull), gTrue, first);
                 lexer = new Lexer(xref, str);
                 parser = new Parser(xref, lexer, gFalse);
                 for (n = 0; n < nObjects; ++n) {
-                    parser->getObj(&obj1);
-                    parser->getObj(&obj2);
+                    obj1 = parser->getObj();
+                    obj2 = parser->getObj();
                     if (n == e->gen)
                         localOffset = obj2.getInt();
-                    obj1.free();
-                    obj2.free();
                 }
 #if defined(POPPLER_VERSION) || defined(XPDF304)
                 while (str->getChar() != EOF) ;
@@ -187,7 +183,6 @@
                 lexer->skipToEOF();
 #endif
                 delete parser;
-                objStr.free();
 
                 fprintf(outfile, "%.10lu 00000 n\n",
                         (long unsigned)(firstOffset + localOffset));
@@ -198,7 +193,6 @@
         s->reset();
         while ((c = s->getChar()) != EOF)
             fputc(c, outfile);
-        srcStream.free();
     }
     if (objnum == 0)
         fprintf(stderr, "Source file extracted to %s\n", outname);
@@ -207,7 +201,6 @@
     else
         fprintf(stderr, "Cross-reference table extracted to %s\n", outname);
     fclose(outfile);
-    catalogDict.free();
     delete doc;
     delete globalParams;
 }
